### ISSUE TITLE
flux-shell: shell plugin design

### ISSUE THREAD

**Issue by grondo** [flux-framework/flux-core#2266]
Both #2165 and #2242 will require a shell plugin design that allows 3rd party code to be registered with the shell and run at different points during job execution.

It would be nice if, similar to Slurm spank plugins, flux-shell plugins could add options to the  `flux {run,srun,jobspec} ...` command lines (e.g. `srun --mpibind`) however perhaps we could get away with a `-o, --option=` generic interface as was used in wreck (not as nice IMO).

Either way, we'll need some way to allow plugins to verify options preferably before the call to submit the jobspec, but worst case as part of the job validation. There's probably many ways to do this, but maybe the job-shell could be run during option parsing with a `--validate` option to validate the generated jobspec *before* it is submitted?

The plugin loader should support loading DSOs in order to work with the new `mpibind` library, but a higher level language would also be useful in the case of the mpi personality plugins. Since the job shell is a user program, we should also support loading user plugins.

For higher level language plugins, Lua is tempting becuase it is trivial to embed in a C program. However, my guess is that most (all?) users would prefer Python. Higher level language support could also be implemented as a C plugin itself (e.g. as in the spank-lua plugin)

**Comment by garlick** [comment]
Can any of the extension module design in [RFC 5](https://github.com/flux-framework/rfc/blob/master/spec_5.adoc) be applied here?

**Comment by grondo** [comment]
We should talk about that. Were you thinking shell plugins should act like broker modules and have a `mod_main` and run with their own reactor? Would the shell need an `rc1` to load these modules? (heh, like a `.flux-shellrc`)

The idea of being able to `flux module load <jobid>.name` is intriguing for instrumenting a running job, though. But I don't think that would be a requirement in the near term and could be added later quite easily actually.

When I was trying to rework the module loading (which I failed at, so take my opinion for what that's worth) I found that conflating broker modules (messaging actors) and plugins for those modules (e.g. sched plugins or in this case shell plugins) in RFC5 *seemed* to unnecessarily constrain the implementation for the plugins case. Maybe I just didn't understand how RFC5 was supposed to apply to the plugin case?

**Comment by garlick** [comment]
I didn't have a fully formed thought there but was thinking, disjointly:
* skimming the python3 embedding documentation, I wondered if JSON encoded "messages" might be an economical way to cross language boundaries, in lieu of direct conversion of data types
* then I wondered if a "message agent" potentially in a separate process could work here
* then I remembered RFC 5 was an attempt to cover "flux plugins" generally.

Maybe RFC 5 should only discuss "actor" modules.  I think the main connection to conventional plugins was that it defined the insmod/rmmod/lsmod RPCs used by the `flux module` front end tool, which could be used to load either type.

**Comment by grondo** [comment]
That's a pretty interesting idea!

It may be quite a bit of work to support though. The shell would have to support "services" for modifying the process state that embedded plugins get you (e.g. `setenv`, `setaffinity`), and you'd have to modify shell code to add new services you forgot about (which negates a reason that plugins may be desirable). Also, there could be some benefit to allowing plugins to use 

If we first build a few plugin hooks into the shell using more "traditional" methods and allow loading of dsos, then a python developer could later come along and make a python loading plugin which used messages to talk to a python script if that was found to be useful.

**Comment by grondo** [comment]
I was thinking we could do something simple for shell plugins at first -- we could create the following hook points in the shell:

 * `init` - called during initialization before any tasks are forked
 * `exit` - called before shell exits
 * `pre_exec` - called from the subprocess pre-exec handler for each task
 * `post_fork` - called from the shell after tasks are forked, before entering reactor (questionable if this one is needed for a user only shell)

The plugin callbacks would have access to an abstract `flux_shell_t *`  which could have a list of getters and setters (maybe getters here could be replaced with a single call that gets all the shell job and/or task information in a single json string?).

This would allow us to easily implement the mpibind plugin at least.

If we feel really strongly that the traditional plugin-callbacks approach isn't right for the flux-shell, we could do something like Slurm did with its "task prolog" scripts (see `--task-prolog` in the [srun](https://slurm.schedmd.com/srun.html.) manpage). The shell could run any configured executable before each task, and communicate with a simple protocol over stdio. This feels a bit more kludgy to me, and probably a lot slower, but avoids plugging directly into shell code if that is the issue here. The recent version of the mpibind script can work as a task prolog with Slurm now.

**Comment by grondo** [comment]
> we could create the following hook points in the shell

I almost forgot:

 - `validate` : called from `flux-shell --validate` mode. Allows a plugin to accept/reject a jobspec.

**Comment by garlick** [comment]
I like it.

So would there be a way for a plugin to provide some functionality to front end tools?  I was wondering if stdio handling could be abstracted in one of these plugins, with one part collecting it among shells, and another part that would allow it to be retrieved by (say) `flux attach`?  Stdio doesn't really need to be a plugin but it would be neat to be able to play with different mechanisms.

**Comment by grondo** [comment]
> So would there be a way for a plugin to provide some functionality to front end tools?

That seems like the tricky question, since we'll have frontend tools written in both C and Python. It would be easy to have plugins have a hook in both the shell and `flux attach` (e.g. similar to spank plugins), but I'm not sure how that would work for a Python `flux run`. Maybe a Python module helper that loads dso plugins?

> I was wondering if stdio handling could be abstracted in one of these plugins, with one part collecting it among shells, and another part that would allow it to be retrieved by (say) flux attach? 

If we added a function to override builtin `shell_io_task_ready()` then probably the existing I/O functionality could be moved to a plugin. We could try an experiment and see what else might need to be exposed to plugins. Since the plugins will have access to the shell's reactor during `init`, and we've already abstracted the addition of new shell services, making arbitrary "plugins" should be possible.

Perhaps the pmi implementation could be made a plugin as well, allowing experimental versions  to be swapped in a at runtime?

**Comment by grondo** [comment]
BTW, I'm still intrigued by the idea of using messages as the abstraction layer for plugins, so we should continue to think about it.

Maybe this is what you already had in mind, but we could still have plugins that run in the same thread of the shell or task being executed (which is necessary or at least convenient when the plugin wants to call `setaffinity` (mpibind) or `setrlimit` (spectrum mpi plugin) or any other self-modifying system call), but could we use a loopback connector to allow these messages to be sent to the same process?

The reason I like this is it does allow some plugins to be run in a separate process (maybe even remotely) if necessary down the road, without a change in interface.

**Comment by garlick** [comment]
We could use either `loop://` (same thread), or `shmem://` (different thread, same process) I guess.
  
With `loop://`, a plugin could safely share `shell->h`; with `shmem://` the shell would need to provide trivial request/response routing for plugins to access broker services.

If we could share plugin loader and routing code in the broker, shell, and potentially front end commands, that would be pretty neat.  Then implement python bindings?

This feels like it might be a bit ambitious for the short term.

**Comment by garlick** [comment]
> Perhaps the pmi implementation could be made a plugin as well, allowing experimental versions to be swapped in a at runtime?

Definitely a good choice for a plugin!    In fact an MPI personality plugin might want to select a PMI implementation optimized for that MPI.

**Comment by garlick** [comment]
Just to get this on a path toward closure (after speaking with @grondo):
* Let's go with the concrete design proposed above by @grondo, or similar
* Conversion of the shell PMI service to a plugin can be the first trial for the new framework

I'll open a new issue on converting PMI to a plugin.

**Comment by grondo** [comment]
I can assign this one to myself and work on an initial implementation, unless someone else is would like to/is already working on it.

In order to support plugin-based I/O modes, we may also want to expose the `shell_task_io_ready` or similar to plugins as well. In summary, the initial design (of the API exported to plugins) I was considering looks something like:

A lot of this is probably subject to change, but a couple design decisions might need to be explained first:

 - This interface minimize the creation of wrappers for existing flux API code by allowing plugins to fetch the underlying shell handle, `flux_cmd_t` (globally and per task), and `flux_subprocess_t`. This allows plugins more power at the expense of requiring plugins be recompiled when Flux API changes (but *not* subject to shell internal implementation at least)
 - Instead of a swath of getters like `flux_shell_get_rank`, `flux_shell_get_jobid` etc, or the more generic `flux_shell_get_item (flux_shell_t *shell, flux_shell_item_t item, ...)` approach, supply plugins with a `get_info` interface that returns a json string encoding the static shell and task information. (inspired by @garlick's comments above) Not sure if it was good idea to include **R* and `jobspec` in this interface (it kind of seems a shame to re-parse jobspec/R in each plugin, so we may want to provide an `unpack` wrapper for these?)

Changes will likely become necessary when actually trying to implement the first plugin using this interface, but htis is just a start.

```C
/************************************************************\
 * Copyright 2019 Lawrence Livermore National Security, LLC
 * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
 *
 * This file is part of the Flux resource manager framework.
 * For details, see https://github.com/flux-framework.
 *
 * SPDX-License-Identifier: LGPL-3.0
\************************************************************/

#ifndef FLUX_SHELL_H
#define FLUX_SHELL_H

#include <flux/core.h>

typedef struct flux_shell      flux_shell_t;
typedef struct flux_shell_task flux_shell_task_t;

typedef int (flux_shell_f)      (flux_shell_t *shell);
typedef int (flux_shell_task_f) (flux_shell_t *shell, flux_shell_task_t *task);
typedef int (flux_shell_io_f)   (flux_shell_t *shell,
                                 flux_shell_task_t *task,
                                 const char *name);

/*  Called from shell's "validate" mode to allow plugins to validate
 *   any options in jobspec
 */
extern flux_shell_f flux_shell_validate;

/*  Called at shell initialization before tasks are launched */
extern flux_shell_f flux_shell_init;

/*  Called just before shell exits */
extern flux_shell_f flux_shell_exit;

/*  Called in each task after fork(2) and before exec(2) */
extern flux_shell_task_f flux_shell_task_init;

/*  Called in the parent shell for each task that is forked */
extern flux_shell_task_f flux_shell_task_post_fork;

/*  Called once for each task after exit */
extern flux_shell_task_f flux_shell_task_exit;


int flux_shell_aux_set (flux_shell_t *shell,
                        const char *key, void *value,
                        flux_free_f freef);

void *flux_shell_aux_get (flux_shell_t *shell, char *key);


flux_t *flux_shell_get_flux (flux_shell_t *shell);
flux_reactor_t *flux_shell_get_reactor (flux_shell_t *shell);
flux_cmd_t *flux_shell_get_cmd (flux_shell_t *shell);

/* info */

/*  Return all shell info as json string:
 *  {
 *   "jobid":I,
 *   "rank":i,
 *   "size":i,
 *   "options": { "verbose":b, "standalone":b }
 *   "jobspec":o,
 *   "R":o
 *  }
 */
int flux_shell_get_info (flux_shell_t *shell, const char **json_str);

/*
 *  Take a "completion reference" on the shell object `shell`.
 *  This function takes a named reference on the shell so that it will
 *  not consider a job "complete" until the reference is released with
 *  flux_shell_remove_completion_ref().
 *
 *  Returns the reference count for the particular name, or -1 on error.
 */
int flux_shell_add_completion_ref (flux_shell_t *shell,
                                   const char *fmt, ...);

/*
 *  Remove a named "completion reference" for the shell object `shell`.
 *  Once all references have been removed, the shells reactor is stopped
 *  with flux_reactor_stop (shell->r).
 *
 *  Returns 0 on success, -1 on failure.
 */
int flux_shell_remove_completion_ref (flux_shell_t *shell,
                                      const char *fmt, ...);

/*  Send signal `sig` to all currently running tasks in shell.
 */
void flux_shell_killall (flux_shell_t *shell, int sig);


/*  Add an event handler for shell event `subtopic`. Message handler
 *  will be auto-destroyed on flux handle destruction at shell shutdown.
 */
int flux_shell_add_event_handler (flux_shell_t *shell,
                                  const char *subtopic,
                                  flux_msg_handler_f cb,
                                  void *arg);


/*  Register service handler for `method` with the shell.
 */
int flux_shell_service_register (flux_shell_t *shell,
                                 const char *method,
                                 flux_msg_handler_f cb,
                                 void *arg);

/*  Send an rpc to shell `method` by shell rank
 */
flux_future_t *flux_shell_rpc_pack (flux_shell_t *shell,
                                    const char *method,
                                    int shell_rank,
                                    int flags,
                                    const char *fmt, ...);

/*  flux_shell_task_t API */

flux_cmd_t * flux_shell_task_get_cmd (flux_shell_task_t *task);

flux_subprocess_t *flux_shell_task_get_subprocess (flux_shell_task_t *task);

/*  Get task general information as JSON string:
 *  {
 *    "global": { "rank":i, "size":i },
 *    "local":  { "rank":i, "size":i }
 *  }
 */
int flux_shell_task_get_info (flux_shell_task_t *task, const char **json_str);

/*  Call ready_cb when channel `name` is ready for reading.
 *
 *  ready_cb can then use `flux_shell_task_get_subprocess()` and use
 *   flux_subprocess_read() or getline() on the result to get
 *   available data.
 */
int flux_shell_task_channel_subscribe (flux_shell_task_t *shell,
                                       const char *name,
                                       flux_shell_io_f ready_cb,
                                       void *arg);
#endif /* !FLUX_SHELL_H */
```

**Comment by grondo** [comment]
It occurred to me the I/O strategy for the interface above might need a bit more explanation. A plugin could add a new channel for a task by using `flux_cmd_add_channel()` in the `task_init` callback, but `libsubprocess` only allows a single callback for all non-stderr/out channels. So the shell will have to add a callback for all channels that multiplexes all channel output to the appropriate plugin registered callback. Plugins notify the shell of these callbacks with `flux_shell_task_channel_subscribe()`.

For this interface, we treat *all* channels the same, including stderr/out which can be subscribed to by a plugin via their names.

**Comment by garlick** [comment]
This sounds like a great start.

Should the json be versioned, e.g. `flux_shell_get_info_v1()` or similar?

I found I needed the whole program's task mapping down in pmi.c to generate PMI_process_mapping.  In the non-plugin version, that entailed iterating over the `struct rcalc_rankinfo` for each shell.   Maybe `flux_shell_task_get_info()` could take a shell rank parameter?

**Comment by grondo** [comment]
> Should the json be versioned, e.g. flux_shell_get_info_v1() or similar?

Is that so that when the json object changes we add a new function `flux_shell_get_info_v2()`? Will we end up with potentially a lot of these functions? Would it be sufficient to add a `"version"` field to the returned JSON string? Caller could check the version if necessary, or heck just check for existence of the key of interest?

> I found I needed the whole program's task mapping down in pmi.c to generate PMI_process_mapping. In the non-plugin version, that entailed iterating over the struct rcalc_rankinfo for each shell. Maybe flux_shell_task_get_info() could take a shell rank parameter?

`flux_shell_task_get_info()` takes a `flux_shell_task_t *`, however we do  need some way to provide the shell's process mapping to plugins. Maybe a `flux_shell_get_rank_info (flux_shell_t *shell, int rank, char **json_str)` as I think you implied above, and return the contents of the `rcalc_rankinfo` in the shell_info JSON object? I guess another idea would be return the mapping for all ranks as a single JSON object. That information (which task ranks ran where) might actually be useful in the permanent job record.

**Comment by garlick** [comment]
> Is that so that when the json object changes we add a new function flux_shell_get_info_v2()? Will we end up with potentially a lot of these functions? Would it be sufficient to add a "version" field to the returned JSON string? Caller could check the version if necessary, or heck just check for existence of the key of interest?

Ah maybe a non-problem then.   Maybe  what we need to pass over this interface to out of tree plugins is sufficiently contained that we can trivially avoid creating issues with incompatible JSON.

> I guess another idea would be return the mapping for all ranks as a single JSON object. That information (which task ranks ran where) might actually be useful in the permanent job record.

Yeah that might be a good thing, possibly in some compressed form.

**Comment by grondo** [comment]
> Ah maybe a non-problem then. Maybe what we need to pass over this interface to out of tree plugins is sufficiently contained that we can trivially avoid creating issues with incompatible JSON.

I guess the only problem here would be if we changed the content of a key in the returned JSON without changing the key name, this could break callers that were originally developed against an older version of the interface. New keys could be added without breaking existing plugins, and even if a key was removed, an older plugin would presumably fail in a graceful way if an expected key is not found. Still, it might be useful to add a `version` key to the returned object since it seems like good practice.

The main drawback to using JSON here is that every plugin in the stack will have re-parse the same JSON string. We could consider falling back to tried and true methods of individual accessors for shell/task information data, or `flux_shell_get_item` style extensible accessor. A python loading plugin could use these interfaces to create the JSON strings to hand to python-based plugins...

**Comment by grondo** [comment]
In #2271 it was discussed that shell plugins may want to start asynchronous work, and may optionally want to ensure this work completes before the BASS drops (Big Ass Shell Synchronization, aka shell barrier).

One way to offer this capability to plugins would be a function that takes  an "initialization reference" similar to a the completion references already used to exit the reactor. If a plugin engages in asynchronous initialization, it could take one of these references and release it when initialization is complete. Once all references are released, the shell would enter the barrier and continue with startup.

Shell initialization would then be changed to something like:

 * call all plugin `init` callbacks
 * call `flux_reactor_run()`
 * exit reactor once all init references have been released
 * barrier
 * start tasks
 * enter reactor again
 * ...

Presumably if a plugin fails initialzation. there will be a way to exit the barrier with an error (logged to job's stderr)

**Comment by grondo** [comment]
> > I guess another idea would be return the mapping for all ranks as a single JSON object. That information (which task ranks ran where) might actually be useful in the permanent job record.
>
> Yeah that might be a good thing, possibly in some compressed form.

Coming back around to this, I can't come up with a compact task id mapping for jobs that would be useful for shell plugins. Of course, for the PMI case, a simple array of assigned task counts may be sufficient, and quite compact.

But for now we can maybe punt and just go with @garlick's suggestion to add a `flux_shell_get_rank_info()` which takes a `shell_rank` parameter, and returns at
first very basic info for each shell rank:
```c
/*  Return per rank job shell info as json string:
 *
 *  {
 *   "broker_rank":i,
 *   "ntasks":i
 *  }
 */

int flux_shell_get_rank_info (flux_shell_t *shell,
                              int shell_rank,
                              const char **json_str);
```

Eventually we could also add a "resources" object to the JSON returned from this function, perhaps with a flat list of resources assigned to the shell rank, e.g.
```json
{ "cores": "0-1", "gpu": "0" }
```

Which may be useful for an affinity/gpu plugin. Since we don't need this yet, I say we keep the JSON as simple as possible for now?

**Comment by grondo** [comment]
Above, we (I) didn't think through how the shell is going to discover or be told which plugins to load. There do seem to be a couple classes of plugins that we might want to handle differently:
  1. basic plugins which extend shell functionality and are enabled/disabled by options in the jobspec. These can all be loaded in the shell at the same time and coexist.
  2. plugins which offer a core functionality, e.g. PMI, and which a user may want to replace with another plugin.

I'm not sure the best way to handle case 2 above. One idea I was toying with would be to allow an optional "service" attribute for plugins, indicating they add a specific service to the shell.  In the case of plugins offering a "service" the last loaded plugin overrides any existing service name with its symbols (e.g. a user could load a different PMI implementation which would somehow disable the default built-in PMI plugin). Perhaps this functionality could be added after the first cut of flux-shell plugins support, but I wanted to bring up the idea now since I had forgotten about this complexity at first.

We also have a broader issue of how a user might customize the plugins loaded by a shell for each of their jobs. Ideally this would be via  options in the jobspec, which overrides any system and user defaults. At the very least, the shell needs some kind of configuration it reads from a system and user path.

Following the example of the system shell, would it be useful to support "rc" files for the shell? `flux-shell` could first read a system rc file from `/etc/flux/shellrc` or similar, then a per user shell `~/.flux/shellrc`, unless overridden by a jobspec attribute? At first, the rc script/config file would just support loading plugins, but the language/config file could be extended in the future to support changing shell behavior in other ways.

**Comment by garlick** [comment]
The rc way of thinking about this seems very natural to me!

**Comment by grondo** [comment]
This probably should have been closed by #2376. Feel free to reopen if there is more to do here.
